File: Depth-First Search/Binary tree/Find paths/LET_no543_Diameter of Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        res, _ = self.dfs(root)
        return res
    
    def dfs(self, cur):
        if not cur:
            return 0, 0
        
        leftRes, leftMax = self.dfs(cur.left)
        rightRes, rightMax = self.dfs(cur.right)
        
        return max(leftRes, rightRes, leftMax + rightMax), max(leftMax, rightMax) + 1

File: Depth-First Search/Binary tree/Find paths/LET_no543_Diameter of Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int res;
public:
    int diameterOfBinaryTree(TreeNode* root) {
        dfs(root);
        return res;
    }
    int dfs(TreeNode* cur) {
        if (!cur)
            return 0;
        
        int leftMax = dfs(cur->left), rightMax = dfs(cur->right);
        res = max(res, leftMax + rightMax);
        
        return max(leftMax, rightMax) + 1;
    }
};